.github/workflows/workflow.yml

```diff
@@ -13,42 +13,123 @@ on:
       - '**.md'
 
 jobs:
-  test:
-    name: Test on ${{ matrix.os }}
-
+  # Build and unit test
+  build:
     strategy:
       matrix:
         os: [ubuntu-latest, windows-latest, macOS-latest]
       fail-fast: false
-
     runs-on: ${{ matrix.os }}
-
     steps:
-    - uses: actions/checkout@v1
-
-    - uses: actions/setup-node@v1
+    - name: Checkout
+      uses: actions/checkout@v2
+    - name: Setup Node.js
+      uses: actions/setup-node@v1
       with:
         node-version: '12.x'
-
-    - name: Get npm cache directory
+    - name: Determine npm cache directory
       id: npm-cache
       run: |
         echo "::set-output name=dir::$(npm config get cache)"
-
-    - uses: actions/cache@v1
+    - name: Restore npm cache
+      uses: actions/cache@v1
       with:
         path: ${{ steps.npm-cache.outputs.dir }}
         key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
         restore-keys: |
           ${{ runner.os }}-node-
-
     - run: npm ci
-
     - name: Prettier Format Check
       run: npm run format-check
-
     - name: ESLint Check
       run: npm run lint
-
     - name: Build & Test
       run: npm run test
+
+  # End to end save and restore
+  test-save:
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest, macOS-latest]
+    runs-on: ${{ matrix.os }}
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v2
+    - name: Generate files
+      if: runner.os != 'windows'
+      run: __tests__/create-cache-files.sh ${{ runner.os }}
+    - name: Generate files (Windows)
+      if: runner.os == 'windows'
+      run: __tests__/create-cache-files.ps1 ${{ runner.os }}
+    - name: Save cache
+      uses: ./
+      with:
+        key: test-${{ runner.os }}-${{ github.run_id }}
+        path: test-cache
+  test-restore:
+    needs: test-save
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest, macOS-latest]
+    runs-on: ${{ matrix.os }}
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v2
+    - name: Restore cache
+      uses: ./
+      with:
+        key: test-${{ runner.os }}-${{ github.run_id }}
+        path: test-cache
+    - name: Verify cache
+      if: runner.os != 'windows'
```

Comment:
Could we use the same shell script for all OS and specify `shell: bash`?

Reply:
checking...

Reply:
Works fine for this purpose.

I'm generally hesitant to go that route. I've run into issues in the past because env vars contain Windows style paths and doesnt work depending on combination of things (iirc git bash commands vs wsl bash commands)